File: get_cpu_power_usage.zig
const std = @import("std");

const CPU_INPUT_FILE = "/sys/class/hwmon/hwmon7/energy17_input";

fn getCpuEnergy() u64 {
    var file = std.fs.cwd().openFile(CPU_INPUT_FILE, .{}) catch {
        std.debug.print("Failed to open file\n", .{});
        std.process.exit(1);
    };

    var buffer: [64]u8 = undefined;
    const bytes_read = file.read(&buffer) catch {
        std.debug.print("Failed to read file\n", .{});
        std.process.exit(1);
    };

    const s = std.mem.trim(u8, buffer[0..bytes_read], " \t\n\r");

    const energy = std.fmt.parseInt(u64, s, 10) catch {
        std.debug.print("Failed to parse energy value\n", .{});
        std.process.exit(1);
    };
    return energy;
}

pub fn main() !void {
    const initial_energy = getCpuEnergy();
    std.time.sleep(10 * 1_000_000);
    const final_energy = getCpuEnergy();
    const energy_diff: f64 = @floatFromInt(final_energy - initial_energy);
    const energy_diff_joules: f64 = energy_diff / 1000000.0;
    const power_usage: f64 = energy_diff_joules / 0.01;
    std.debug.print("{d:.1}\n", .{power_usage});
}

File: get_cpu_power_usage.rs
use std::fs::read_to_string;
use std::process;
use std::thread::sleep;
use std::time::Duration;

const CPU_INPUT_FILE: &str = "/sys/class/hwmon/hwmon7/energy17_input";

fn get_cpu_energy() -> u64 {
    let content = read_to_string(CPU_INPUT_FILE).unwrap_or_else(|err| {
        eprintln!("Failed to read file: {}", err);
        process::exit(1);
    });
    let energy = content.trim().parse::<u64>().unwrap_or_else(|err| {
        eprintln!("Failed to parse energy value: {}", err);
        process::exit(1);
    });
    energy
}

fn main() {
    let initial_energy = get_cpu_energy();
    sleep(Duration::from_millis(10));
    let final_energy = get_cpu_energy();
    let energy_diff = final_energy - initial_energy;
    let energy_diff_joules = energy_diff as f64 / 1e6;
    let power_usage = energy_diff_joules / 0.01;
    println!("{:.1}", power_usage);
}
